test/hpke_test.c

```diff
@@ -0,0 +1,1801 @@
+        { fourth_context3, sizeof(fourth_context3),
+          fourth_export3, sizeof(fourth_export3) },
+    };
+    return do_testhpke(&basedata, NULL, 0,
+                       exportdata, OSSL_NELEM(exportdata));
+}
+
+/*
+ * Randomly toss a coin
+ */
+#define COIN_IS_HEADS (test_random() % 2)
+
+/* tables of HPKE modes and suite values */
+static int hpke_mode_list[] = {
+    OSSL_HPKE_MODE_BASE,
+    OSSL_HPKE_MODE_PSK,
+    OSSL_HPKE_MODE_AUTH,
+    OSSL_HPKE_MODE_PSKAUTH
+};
+static uint16_t hpke_kem_list[] = {
+    OSSL_HPKE_KEM_ID_P256,
+    OSSL_HPKE_KEM_ID_P384,
+    OSSL_HPKE_KEM_ID_P521,
+    OSSL_HPKE_KEM_ID_X25519,
+    OSSL_HPKE_KEM_ID_X448
+};
+static uint16_t hpke_kdf_list[] = {
+    OSSL_HPKE_KDF_ID_HKDF_SHA256,
+    OSSL_HPKE_KDF_ID_HKDF_SHA384,
+    OSSL_HPKE_KDF_ID_HKDF_SHA512
+};
+static uint16_t hpke_aead_list[] = {
+    OSSL_HPKE_AEAD_ID_AES_GCM_128,
+    OSSL_HPKE_AEAD_ID_AES_GCM_256,
+    OSSL_HPKE_AEAD_ID_CHACHA_POLY1305
+};
+
+/*
+ * Strings that can be used with names or IANA codepoints.
+ * Note that the initial entries from these lists should
+ * match the lists above, i.e. kem_str_list[0] and
+ * hpke_kem_list[0] should refer to the same KEM. We use
+ * that for verbose output via TEST_note() below.
+ * Subsequent entries are only used for tests of
+ * OSSL_HPKE_str2suite()
+ */
+static char *mode_str_list[] = {
+    "base", "psk", "auth", "pskauth"
+};
+static char *kem_str_list[] = {
+    "P-256", "P-384", "P-521", "x25519", "x448",
+    "0x10", "0x11", "0x12", "0x20", "0x21",
+    "16", "17", "18", "32", "33"
+};
+static char *kdf_str_list[] = {
+    "hkdf-sha256", "hkdf-sha384", "hkdf-sha512",
+    "0x1", "0x01", "0x2", "0x02", "0x3", "0x03",
+    "1", "2", "3"
+};
+static char *aead_str_list[] = {
+    "aes-128-gcm", "aes-256-gcm", "chacha20-poly1305", "exporter",
+    "0x1", "0x01", "0x2", "0x02", "0x3", "0x03",
+    "1", "2", "3",
+    "0xff", "255"
+};
+/* table of bogus strings that better not work */
+static char *bogus_suite_strs[] = {
+    "3,33,3",
+    "bogus,bogus,bogus",
+    "bogus,33,3,1,bogus",
+    "bogus,33,3,1",
+    "bogus,bogus",
+    "bogus",
+    /* and one in the wrong order */
+    "aes-256-gcm,hkdf-sha512,x25519"
+};
+
+/**
+ * @brief round-trips, generating keys, encrypt and decrypt
+ *
+ * This iterates over all mode and ciphersuite options trying
+ * a key gen, encrypt and decrypt for each. The aad, info, and
+ * seq inputs are randomly set or omitted each time. EVP and
+ * non-EVP key generation are randomly selected.
+ *
+ * @return 1 for success, other otherwise
+ */
+static int test_hpke_modes_suites(void)
+{
+    int overallresult = 1;
+    int mind = 0; /* index into hpke_mode_list */
+    int kemind = 0; /* index into hpke_kem_list */
+    int kdfind = 0; /* index into hpke_kdf_list */
+    int aeadind = 0; /* index into hpke_aead_list */
+
+    /* iterate over the different modes */
+    for (mind = 0; mind != OSSL_NELEM(hpke_mode_list); mind++) {
```

Comment:
you better split this test using `ADD_ALL_TESTS` fixture to iterate on `hpke_mode_list` ...

so it will split a bit the test set, adding a finer granularity to test progression, and so ease any debugging.

Reply:
I had thought about doing so but I'd argue that's not needed as you can specify verbose output if one of these combinations fail, i.e. ``test/hpke_test -v`` will produce output like that below and will stop on the first failure. (It's also the case that, so far, there're no applications of HPKE I know of that use other than the base mode;-)

```
$ test/hpke_test -test 6 -v 
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha256/0x01, aead: aes-128-gcm/0x01
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha256/0x01, aead: aes-256-gcm/0x02
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha256/0x01, aead: chacha20-poly1305/0x03
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha384/0x02, aead: aes-128-gcm/0x01
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha384/0x02, aead: aes-256-gcm/0x02
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha384/0x02, aead: chacha20-poly1305/0x03
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha512/0x03, aead: aes-128-gcm/0x01
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha512/0x03, aead: aes-256-gcm/0x02
# HPKE worked for mode: base/0x00, kem: P-256/0x10, kdf: hkdf-sha512/0x03, aead: chacha20-poly1305/0x03
# HPKE worked for mode: base/0x00, kem: P-384/0x11, kdf: hkdf-sha256/0x01, aead: aes-128-gcm/0x01
# HPKE worked for mode: base/0x00, kem: P-384/0x11, kdf: hkdf-sha256/0x01, aead: aes-256-gcm/0x02
...
```